app/src/org/commcare/android/nfc/NdefRecordUtil.java

```diff
@@ -0,0 +1,133 @@
+        }
+    }
+
+    private static Pair<String,Boolean> handleExternalTypeRecord(NdefRecord record,
+                                                                 String[] userSpecifiedTypes,
+                                                                 String domainForExternalTypes) {
+        String typeAsString = new String(record.getType(), UTF8_CHARSET);
+        if (recordTypeIsExpected(typeAsString, userSpecifiedTypes, domainForExternalTypes)) {
+            return new Pair<>(new String(record.getPayload(), UTF8_CHARSET), true);
+        } else {
+            return new Pair<>(READ_ERROR_MISMATCH_KEY, false);
+        }
+    }
+
+    private static boolean recordTypeIsExpected(String recordTypeAsString,
+                                                String[] userSpecifiedTypes,
+                                                String domainForExternalTypes) {
+        for (String type : userSpecifiedTypes) {
+            type = domainForExternalTypes + ":" + type;
+            if (type.equals(recordTypeAsString)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private static String readValueFromTextRecord(NdefRecord textTypeRecord) {
+        byte[] fullPayload = textTypeRecord.getPayload();
+        // The payload includes a prefix denoting the language, so we need to parse that off
+        int langBytesLength = fullPayload[0]; // status byte
+        int lengthOfPrefix = langBytesLength + 1; // add 1 for the status byte itself
+        byte[] payloadWithoutLang = Arrays.copyOfRange(fullPayload, lengthOfPrefix, fullPayload.length);
+        return new String(payloadWithoutLang, UTF8_CHARSET);
+    }
+
+    protected static NdefRecord createNdefRecord(String userSpecifiedType,
+                                                 String userSpecifiedDomain,
+                                                 String payloadToWrite) {
+        if (isCommCareSupportedWellKnownType(userSpecifiedType)) {
+            return createWellKnownTypeRecord(userSpecifiedType, payloadToWrite);
+        } else {
+            return createExternalRecord(userSpecifiedType, userSpecifiedDomain, payloadToWrite);
+        }
+    }
+
+    protected static boolean isCommCareSupportedWellKnownType(String type) {
+        // For now, the only "well known type" we're supporting is NdefRecord.RTD_TEXT, which
+        // users should encode in their configuration by specifying type "text"
+        return "text".equals(type);
+    }
+
+    private static NdefRecord createWellKnownTypeRecord(String type, String payload) {
+        if (type.equals("text")) {
+            return createTextRecord(payload);
+        } else {
+            throw new IllegalArgumentException();
+        }
+    }
+
+    private static NdefRecord createTextRecord(String payload) {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+            return NdefRecord.createTextRecord(null, payload);
```

Comment:
I don't know how to evaluate whether they are "identical" per say, but I have tested that both paths create an Ndef Record of type text that is readable by both CommCare and the Android device's default NFC processing system. So that seems sufficient to me. Android's NFC implementation guidelines strongly suggest using their built-in NdefRecord create methods where possible, and we also follow this sort of pattern quite often (where we follow 2 different paths depending on Android version so as to use the newer/better standard where possible).